content/blog/towards-continuous-performance-regression-testing.asciidoc

```diff
@@ -0,0 +1,594 @@
+---
+title: "Towards Continuous Performance Regression Testing"
+date: 2020-12-16T17:45:00+01:00
+draft: false
+---
+:source-highlighter: rouge
+:rouge-style: base16.dark
+:icons: font
+:imagesdir: /images
+ifdef::env-github[]
+:imagesdir: ../../static/images
+endif::[]
+
+Functional unit and integration tests are a standard tool of any software development organization,
+helping not only to ensure correctness of newly implemented code,
+but also to identify regressions -- bugs to existing functionality introduced by a code change.
+The situation looks different though when it comes to regressions related to non-functional requirements, in particular performance-related ones:
+How to detect increased response times in a web application?
+How to identify decreased throughput?
+
+These aspects are typically hard to test in an automated and reliable way in the development workflow,
+as they are dependent on the underlying hardware and the workload of an application.
+For instance assertions on the duration of specific requests of a web application cannot be run in a meaningful way on a developer laptop which differs from the actual production hardware.
+When run in a virtualized or containerized CI environment, such tests are prone to failures due to concurrent load of other applications and jobs.
```

Comment:
I'm assuming you are suggesting the idea of running performance tests on virtualized / shared resources here, for example we have some tests which send an alert if some metric gets beyond some % trheshold of difference with previous averages. Might want to clarify that?

In which case I'd say s/failures/ severe measurement distortions / ?

There's also other challenges: valid bugfixes and good improvements which tend to slowly get you closer to thethreshold of failure will regularly erode safety margins, eventually triggering CI failures for totally innocent changes; for example I've seen a typo in a README trigger a performance failure - because such metrics need constant "gardening" to be useful.

Reply:
> I'd say s/failures/ severe measurement distortions / ?

+1 I like that.

> There's also other challenges [...]

Yes, great point. Added this to the section at the end about possible future additions. Also discussing with John O'Hara about this who has similar ideas.